pyproject.toml

```diff
@@ -1,6 +1,6 @@
 [build-system]
 requires = [
-    "Cython >=0.29.31,<1",
+    "Cython >=3",
```

Comment:
Does this actually no longer work with Cython 3.0, or did you just change this to ensure it's using a newer version on CI?

I don't think it's a problem if we require a recent version of Cython, but I'd like to check that if it really doesn't work on Cython 0.x any more.

Reply:
I think it is fine on older Cython's for the syntax used in this PR. I've been doing this change in my own projects to avoid inconsistent builds. I haven't seen any major troubles between different builds beyond some Cython 3 bugs that were fixed. Just a little scared to let it choose whatever. Not a strong opinion on my part as a non-maintainer of this project.

Reply:
Oh! Maybe this should set a limit (not sure what to) since Cython is deprecating the `IF` compiler condition (or whatever it is called in Cython-land). I'm not sure if a version has been set when it will actually be removed, but I also know people are still working on the best suggestions for migrating away from it (literal C code, macros, etc).

Reply:
Good point. Let's put an upper bound of `<4` on it, in the hope that actually removing `IF` would trigger another major version. Last I heard, the warnings about it were being turned off for now while migration paths are worked out, so I don't think removal is imminent :crossed_fingers: .

And if the syntax still works on 0.29.x, let's leave the lower bound as it is for now. Although I don't think we actually test building with older Cythons, so it may break without us noticing. :shrug: Testing at the intersection of so many big dependencies (HDF5, Numpy, Python, Cython) is kind of a headache.

Reply:
> And if the syntax still works on 0.29.x, let's leave the lower bound as it is for now

Leave it as it is in `master` or leave it as it is in this PR? I'll assume 0.29 included.